packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,417 @@
+            const { log } = state;
+
+            return log.dump();
+          },
+          getOptFatalProblem() {
+            const { facets } = this;
+            const { flow } = facets;
+
+            return failures.has(flow) ? failures.get(flow) : undefined;
+          },
+        },
+        admin: {
+          reset() {
+            const { state, facets } = this;
+            const { bijection, log } = state;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.delete(flow);
+            }
+            if (eagerWakers.has(flow)) {
+              // For now, once an eagerWaker, always an eagerWaker
+              // eagerWakers.delete(flow);
+            }
+            if (eph.membrane) {
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            state.isDone = false;
+          },
+          done() {
+            const { state, facets } = this;
+            const { log } = state;
+            const { admin } = facets;
+
+            admin.reset();
+            state.isDone = true;
+            log.dispose();
+          },
+          panic(fatalProblem) {
+            const { state, facets } = this;
+            const { bijection, log } = state;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.set(flow, fatalProblem);
+            } else {
+              failures.init(flow, fatalProblem);
+            }
+
+            if (eph.membrane) {
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            // This is not an expected throw, so in theory arbitrary chaos
+            // may ensue from throwing it. But at this point
+            // we should have successfully isolated this activation from
+            // having any observable effects on the host, aside from
+            // console logging and
+            // resource exhaustion, including infinite loops
+            const err = makeError(
+              X`In a replay failure: see getFailures() for more information`,
+            );
+            annotateError(err, X`due to ${fatalProblem}`);
+            throw err;
+          },
+        },
+        wakeWatcher: {
+          onFulfilled(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+          onRejected(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+        },
+      },
+    );
+    const makeAsyncFlowKit = (activationThis, activationArgs) => {
+      const asyncFlowKit = internalMakeAsyncFlowKit(
+        activationThis,
+        activationArgs,
+      );
+      asyncFlowKit.flow.restart();
+      return asyncFlowKit;
+    };
```

Comment:
Why is this needed? The caller `asyncFlow` is "internal", right? Or is `prepareAsyncFlowKit` meant to be used outside of tests?

Reply:
> Or is `prepareAsyncFlowKit` meant to be used outside of tests?

Yes. The `asyncFlow` function is a convenience we expect to account for typical usage. But it weirdly breaks our normal `prepareFoo` patterns. For those wishing to more directly abstract of async flow --- like perhaps orchestration! --- we also expose the normal `prepareFoo` APIs.

Resolving.